### PR TITLE
172747567 cost function

### PR DESCRIPTION
**Pull request by mstechly**

### CODE REVIEW

File: src/python/zquantum/optimizers/scipy_optimizer.py [hunk 1]
```diff
@@ -47,4 +47,6 @@ def default_callback(params):
         result.history = history
         if 'hess_inv' in result.keys():
             del result['hess_inv']
+        if 'final_simplex' in result.keys():
+            del result['final_simplex']
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/scipy_optimizer.py:51]: How come this is coming about now and wasn't a problem in the past?

File: src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py [hunk 2]
```diff
@@ -1,10 +1,8 @@
...
 from zquantum.core.utils import load_value_estimate, convert_array_to_dict, SCHEMA_VERSION
-
+from .cost_function import ProxyCostFunction
 import scipy
-import time
 import json
-import io
 
 # ---------- Optimize Variational Circuit ----------
 def optimize_variational_circuit_with_proxy(initial_params, optimizer, client, **kwargs):
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py:8]: I think we might be able to collapse this into the `optimize_variational_circuit` function by making the initial and final status calls to the proxy in the template.

File: templates/daemon-optimizer.yaml [hunk 3]
```diff
@@ -27,8 +27,8 @@ spec:
         path: /app/start_proxy.py
         raw:
           data: |
-            from zquantum.core.optimizer.optimize import optimize_variational_circuit_with_proxy
-            from zquantum.core.optimizer.proxy import Client
+            from zquantum.optimizers.daemon_optimizer.optimize import optimize_variational_circuit_with_proxy
```
Comment (simonwa7) [inline templates/daemon-optimizer.yaml:30]: If we can remove `optimize_variational_circuit_with_proxy` then we will need to update this template.

File: src/python/zquantum/optimizers/scipy_optimizer.py (hunk 1 above)
Comment (mstechly) [reply src/python/zquantum/optimizers/scipy_optimizer.py:51]: No idea, I think no one used Nelder-Mead for quite some time :)

File: src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py (hunk 2 above)
Comment (simonwa7) [reply src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py:8]: Apologies -> meant collapse this into intializing the cost function and calling optimizer.minimize in the template

File: src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py [hunk 4]
```diff
@@ -38,65 +36,19 @@ def optimize_variational_circuit_with_proxy(initial_params, optimizer, client, *
 
     opt_results = scipy.optimize.OptimizeResult({'value': None, 'status': None, 'success': False, 'nfev': 0,
         'nit': 0, 'history':[{'optimization-evaluation-ids': []}]})
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py:38]: I believe lines 32 through 38 are all unnecessary given the changes.

File: src/python/zquantum/optimizers/optimizers_test.py [hunk 5]
```diff
@@ -14,15 +15,15 @@ class TestOptimizers(unittest.TestCase):
...
+        self.cost_function = BasicCostFunction(rosen)
         self.scipy_optimizer = ScipyOptimizer(method="L-BFGS-B")
         self.grid_search_optimizer = GridSearchOptimizer(grid)
         self.cmaes_optimizer = CMAESOptimizer(options={"sigma_0": 0.1})
         self.optimizers = [self.scipy_optimizer, self.grid_search_optimizer, self.cmaes_optimizer]
 
     def test_optimization(self):
         for optimizer in self.optimizers:
-            results = optimizer.minimize(rosen, initial_params=[0, 0])
+            results = optimizer.minimize(self.cost_function, initial_params=[0, 0])
```
Comment (simonwa7) [inline src/python/zquantum/optimizers/optimizers_test.py:26]: I believe this should be `self.cost_function.evaluate` -> method for the cost function evaluation, not the `CostFunction` object

Comment (mstechly) [reply src/python/zquantum/optimizers/optimizers_test.py:26]: I disagree.
There might be additional logic inside the `minimize` function for handling, that requires access to the whole `cost_function` object.
E.g. history of evaluations. Also, we might want to modify the cost function during the runtime, for example for the adaptive optimizers.

Comment (simonwa7) [reply src/python/zquantum/optimizers/optimizers_test.py:26]: If this is the case, then the scipy optimizer needs to be updated.

Comment (mstechly) [reply src/python/zquantum/optimizers/optimizers_test.py:26]: Addressed this in: [`13cd5a`](https://github.com/zapatacomputing/z-quantum-optimizers/commit/13cd5a067549bf8f25ba3796ce73be98a31ebc4c)

File: src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py (hunk 4 above)
Comment (mstechly) [reply src/python/zquantum/optimizers/daemon_optimizer/optimize/_api.py]: Addressed this in [`203825`](https://github.com/zapatacomputing/z-quantum-optimizers/commit/203825750c6cc6261fa29aec3331333e475238d3)

**Pull request merged**